File: java/src/com/datastructures/trees/BinarySearchTree.java
package com.datastructures.trees;

import java.util.ArrayList;

public class BinarySearchTree<T extends Comparable<T>> {
    BinarySearchNode<T> head; //root
    int count;
    public BinarySearchTree() {
        count = 0;
    }
    
    public BinarySearchTree(T value) {
       add(value);
    }
    public void add(T value) {
        BinarySearchNode<T> newNode = new BinarySearchNode<T>(value);
        if(head == null) {
            head = newNode;
            return;
        }
        addHelper(head, newNode);
    }

    private void addHelper(BinarySearchNode<T> curr, 
                           BinarySearchNode<T> newNode) {
        if(newNode.compareTo(curr) > 0) {
            if(curr.getRight() == null) {
                curr.setRight(newNode);
                return;
            } else {
                addHelper(curr.getRight(), newNode);
            }
        } else {
            if(curr.getLeft() == null) {
                curr.setLeft(newNode);
                return;
            } else {
                addHelper(curr.getLeft(), newNode);
            }
        }
    }

    public ArrayList<T> getInorder() {
        ArrayList<T> returnList = new ArrayList<T>();
        getInorder(head, returnList);
        return returnList;

    }

    private void getInorder(BinarySearchNode<T> curr, ArrayList<T> arrList) {
        if(curr == null) {
            return;
        }
        getInorder(curr.getLeft(), arrList);
        arrList.add(curr.getData());
        getInorder(curr.getRight(), arrList);
    }

    private int maxDepth(BinarySearchNode<T> curr) {
        if(curr == null) {
            return 0;
        }
        return 1 + Math.max(maxDepth(curr.getLeft()), maxDepth(curr.getRight()));
    }

    private int minDepth(BinarySearchNode<T> curr) {
        if(curr == null) {
            return 0;
        }
        return 1 + Math.min(minDepth(curr.getLeft()), minDepth(curr.getRight()));
    }

    // 4.1 Cracking the coding interview
    public boolean isBalanced() {
        return maxDepth(this.head) - minDepth(this.head) <= 1;
    }


}

File: c#/datastructures/BinarySearchTree.cs
using System;
using System.Collections.Generic;

namespace datastructures
{
    public class BSTNode {
        public BSTNode Left {get; set;}
        public BSTNode Right {get; set;}
        public int data {get; private set;} // TODO make this generic with c# equivalent of comparable
        public int height {get; set;}
        public BSTNode(int data) {
            this.data = data;
        }
    }
    public class BinarySearchTree {
        public BSTNode Root {get; private set;}
        public int count {get; private set;}
        public BinarySearchTree() {
            count = 0;
        }
        
        public void add(int value) {
            BSTNode newLeaf = new BSTNode(value);
            if(Root == null) {
                Root = newLeaf;
                Root.height = 1;
                count = 1;
                return;
            }
            addHelper(Root, newLeaf);
            count++;
        }

        // if newLeaf > curr go right
        // if newLeaf <= curr go left
        private void addHelper(BSTNode curr, BSTNode newLeaf) {
            if (newLeaf.data > curr.data) {
                if(curr.Right == null) {
                    curr.Right = newLeaf;
                    curr.Right.height = curr.height + 1;
                    return;
                } else {
                    addHelper(curr.Right, newLeaf);
                }
            } else {
                if(curr.Left == null) {
                    curr.Left =  newLeaf;
                    curr.Left.height = curr.height + 1;
                    return;
                } else {
                    addHelper(curr.Left, newLeaf);
                }
            }
        }
        public List<int> getInorder() {
            List<int> returnList = new List<int>();
            getInorder(Root, returnList);
            return returnList;
        }

        private void getInorder(BSTNode curr, List<int> arrList) {
            if(curr == null) {
                return;
            }
            getInorder(curr.Left, arrList);
            arrList.Add(curr.data);
            getInorder(curr.Right, arrList);
        }

        public List<int> getPreorder() {
            List<int> returnList = new List<int>();
            getPreorder(Root, returnList);
            return returnList;
        }

        private void getPreorder(BSTNode curr, List<int> arrList) {
            if(curr == null) {
                return;
            }
            arrList.Add(curr.data);
            getPreorder(curr.Left, arrList);
            getPreorder(curr.Right, arrList);
        }

        public List<int> getPostrder() {
            List<int> returnList = new List<int>();
            getPostrder(Root, returnList);
            return returnList;
        }

        private void getPostrder(BSTNode curr, List<int> arrList) {
            if(curr == null) {
                return;
            }
            getPostrder(curr.Left, arrList);
            getPostrder(curr.Right, arrList);
            arrList.Add(curr.data);
        }

        public int maxDepth() {
            return maxDepth(Root);
        }
        /*private int max(int a, int b) { // replace with library
            if (a > b) {
                return a;
            }
            return b;
        }*/
        private int maxDepth(BSTNode curr) {
            if(curr == null) {
                return 0;
            }
            return 1 + Math.Max(maxDepth(curr.Left), maxDepth(curr.Right));
        }

        public bool isBallanced() {
            return isBalanced(Root);
        }

        private bool isBalanced(BSTNode curr) {
            if(curr == null) {
                return true;
            }
            int heightDiff = maxDepth(curr.Left) - maxDepth(curr.Right);
            if (Math.Abs(heightDiff) > 1) {
                return false;
            } else { // recurse
                return isBalanced(curr.Left) && isBalanced(curr.Right);
            }
        }

    }
}
